Unify HID usage page approach to avoid consumer page crazy.

petejohanson opened the pull request:
* Remove need for separate `&cp` behavior, but
  keep it for now for backward compat.
* Refactor sensor inc/dec as well.

okke-formsma commented:
lgtm

innovaker commented on app/include/dt-bindings/zmk/keys.h:
```diff
@@ -6,6 +6,8 @@
 
 #pragma once
 
+#define EXT_USAGE(page, id) ((page << 16) + id)
```
We could also consider turning it into `HID_USAGE_EXT` or `HID_EXT_USAGE` and putting it into `hid_usage.h` along with the HID usage codes perhaps?

petejohanson replied:
Yes, I think putting this ultimately in `hid_usage.h` is the right place for this.

innovaker reviewed (changes requested):
Thanks for doing this @petejohanson!  Looks good overall, just a few suggestions.

I've implemented them in #311 for testing.  I didn't refactor `EXT_USAGE` or the compilation warnings though.

petejohanson commented on app/include/zmk/events/keycode-state-changed.h:
```diff
@@ -18,11 +19,23 @@ struct keycode_state_changed {
+static inline struct keycode_state_changed *keycode_state_changed_from_encoded(u32_t encoded,
+                                                                               bool pressed) {
+    u16_t page = (encoded >> 16) & 0x0F;
+    u16_t id = encoded & 0xFF;
```
Indeed. Late night coding was not kind to me on this PR.

petejohanson commented on app/include/zmk/events/keycode-state-changed.h:
```diff
@@ -18,11 +19,23 @@ struct keycode_state_changed {
+
+static inline struct keycode_state_changed *keycode_state_changed_from_encoded(u32_t encoded,
+                                                                               bool pressed) {
+    u16_t page = (encoded >> 16) & 0x0F;
```
Sure were.

innovaker commented on app/include/dt-bindings/zmk/keys.h:
```diff
@@ -6,6 +6,8 @@
 
 #pragma once
 
+#define EXT_USAGE(page, id) ((page << 16) + id)
```
#153 used the bitwise `OR` notation.  We should be consistent whichever we settle on.

petejohanson commented:
@innovaker I have rebased this, and incorporated just the `keys.h` changes from #311 so to update the defines to incorporate the usage page in the values. Can you take a look and give me your thoughts?

innovaker reviewed (changes requested):
Looking good!

Perhaps check the headers.  I noticed some were obsolete and some were indirect - mostly around `hid_usage_pages.h` and `keys.h`.

I'd personally put the keymap changes into a follow up commit on the same PR.

Will take a closer look and do some testing tomorrow.

petejohanson commented on app/include/zmk/events/keycode-state-changed.h:
```diff
@@ -29,3 +31,21 @@ create_keycode_state_changed(u8_t usage_page, u32_t keycode, bool state) {
+    zmk_mod_flags implicit_mods = SELECT_MODS(encoded);
+
+    if (!page) {
+        page = USAGE_KEYPAD;
```
Fixed.

petejohanson commented on app/include/zmk/events/keycode-state-changed.h:
```diff
@@ -9,6 +9,8 @@
 #include <zephyr.h>
 #include <dt-bindings/zmk/modifiers.h>
 #include <zmk/event-manager.h>
+#include <dt-bindings/zmk/keys.h>
```
Indeed. Fixed and in a better order.

innovaker commented on docs/docs/dev-guide-new-shield.md:
```diff
@@ -393,7 +393,6 @@ The two `#include` lines at the top of the keymap are required in order to bring
 Further documentation on behaviors and bindings is forthcoming, but a summary of the current behaviors you can bind to key positions is as follows:
 
 - `kp` is the "key press" behavior, and takes a single binding argument of the HID keycode from the 'keyboard/keypad" HID usage table.
```
> and takes a single binding argument of the HID keycode from the 'keyboard/keypad" HID usage table.

Does this need generalizing?

petejohanson replied:
I changed the wording a bit to be more accurate, but left the term "keycode" in parentheses to help people understand.

innovaker replied:
Given that this particular document is aimed at people who are unlikely to care about what the code is made of (HID parts), I'd just keep it simple.  Perhaps just `... for key codes` ... that is until we come up with a better work to replace it (something more encompassing that includes other controls).

petejohanson replied:
Updated.

innovaker reviewed (changes requested):
Looking good.  Did a wider check though ...
 - `inc_dec_kp` and `inc_dec_cp` still exists in
    - keymaps
    - documentation.

end-2-end tests passed:
 - [e2e-2020-11-05T09-21-00-ble.log](https://github.com/zmkfirmware/zmk/files/5493353/e2e-2020-11-05T09-21-00-ble.log)
 - [e2e-2020-11-05T09-21-00-usb.log](https://github.com/zmkfirmware/zmk/files/5493354/e2e-2020-11-05T09-21-00-usb.log)

As discussed:
 - `HID_EXT_USAGE` macros might deserve their own header.
 - In the future, we should consider refactoring the system so that `keycode` is unpacked into a struct as early as possible and solely using the unpacked values (page, usage) for internal logic.  This will also reduce the coupling of the keymap encoding to the core logic which will be beneficial as we move away from DT for keymaps.

petejohanson commented:
@innovaker I'm going to hold off on moving the extra macros to their own header for now, They're generally used together, so I think until we have more mixing of concerns they can live together.

Regarding the "decompose sooner", I'm not sure how much sooner we could do it, they are encoded as `u32_t` values in the keymap since the keymap uses that as a generic storage mechanism for a variety of different kinds of parameters. As soon as it hits a behavior that is actually HID/keycode related, it is decoded and stored as discrete values in the event.

Fixed remaining keymaps and docs.

innovaker commented on app/src/behaviors/behavior_hold_tap.c:
```diff
@@ -429,8 +429,8 @@ static int position_state_changed_listener(const struct zmk_event_header *eh) {
 static inline bool only_mods(struct keycode_state_changed *ev) {
-    return ev->usage_page == HID_USAGE_KEY && ev->keycode >= LEFT_CONTROL &&
-           ev->keycode <= RIGHT_GUI;
+    return ev->usage_page == HID_USAGE_KEY && ev->keycode >= HID_EXT_USAGE_ID(LEFT_CONTROL) &&
```
`HID_USAGE_KEY_KEYBOARD_LEFTCONTROL`?

innovaker commented on app/src/behaviors/behavior_hold_tap.c:
```diff
@@ -429,8 +429,8 @@ static int position_state_changed_listener(const struct zmk_event_header *eh) {
-    return ev->usage_page == HID_USAGE_KEY && ev->keycode >= LEFT_CONTROL &&
-           ev->keycode <= RIGHT_GUI;
+    return ev->usage_page == HID_USAGE_KEY && ev->keycode >= HID_EXT_USAGE_ID(LEFT_CONTROL) &&
+           ev->keycode <= HID_EXT_USAGE_ID(RIGHT_GUI);
```
`HID_USAGE_KEY_KEYBOARD_RIGHT_GUI`?

innovaker commented on app/include/zmk/events/keycode-state-changed.h:
```diff
@@ -20,12 +22,20 @@ struct keycode_state_changed {
+static inline struct keycode_state_changed *keycode_state_changed_from_encoded(u32_t encoded,
+                                                                               bool pressed) {
+    u16_t page = (encoded >> 16) & 0xFF;
+    u16_t id = encoded & 0xFFFF;
```
`HID_EXT_USAGE_ID(...)`?

innovaker commented on app/include/zmk/events/keycode-state-changed.h:
```diff
@@ -20,12 +22,20 @@ struct keycode_state_changed {
-create_keycode_state_changed(u8_t usage_page, u32_t keycode, bool state) {
+static inline struct keycode_state_changed *keycode_state_changed_from_encoded(u32_t encoded,
+                                                                               bool pressed) {
+    u16_t page = (encoded >> 16) & 0xFF;
```
`HID_EXT_USAGE_PAGE(...)`?

innovaker reviewed (changes requested):
Will do a proper review after food, but just spotted the above and figured I should point them out sooner rather than later.  We should probably be consistent in how we deconstruct and also use the `HID_USAGE_` constants where appropriate.

petejohanson commented:
@innovaker Good catch. Also fixed in `hid.c` while I was at it.

innovaker commented on app/src/behaviors/behavior_hold_tap.c:
```diff
@@ -429,8 +429,9 @@ static int position_state_changed_listener(const struct zmk_event_header *eh) {
-    return ev->usage_page == HID_USAGE_KEY && ev->keycode >= LEFT_CONTROL &&
-           ev->keycode <= RIGHT_GUI;
+    return ev->usage_page == HID_USAGE_KEY &&
+           ev->keycode >= HID_EXT_USAGE_ID(HID_USAGE_KEY_KEYBOARD_LEFTCONTROL) &&
+           ev->keycode <= HID_EXT_USAGE_ID(HID_USAGE_KEY_KEYBOARD_RIGHT_GUI);
```
Isn't applying `HID_EXT_USAGE_ID(...)` now superfluous?

innovaker commented on app/src/hid.c:
```diff
@@ -78,16 +79,18 @@ int zmk_hid_implicit_modifiers_release() {
 int zmk_hid_keypad_press(zmk_key code) {
-    if (code >= LEFT_CONTROL && code <= RIGHT_GUI) {
-        return zmk_hid_register_mod(code - LEFT_CONTROL);
+    if (code >= HID_EXT_USAGE_ID(HID_USAGE_KEY_KEYBOARD_LEFTCONTROL) &&
+        code <= HID_EXT_USAGE_ID(HID_USAGE_KEY_KEYBOARD_RIGHT_GUI)) {
+        return zmk_hid_register_mod(code - HID_EXT_USAGE_ID(HID_USAGE_KEY_KEYBOARD_LEFTCONTROL));
```
`HID_EXT_USAGE_ID` superfluous?

petejohanson replied:
No, the code is *just* the usage ID, not the full encoded value, so if you did that, you'd subtract the high bits of the page as well, which is not intended.

petejohanson replied:
Oh! I'm sorry, yes, this is.

innovaker commented on app/src/hid.c:
```diff
@@ -78,16 +79,18 @@ int zmk_hid_implicit_modifiers_release() {
 int zmk_hid_keypad_release(zmk_key code) {
-    if (code >= LEFT_CONTROL && code <= RIGHT_GUI) {
-        return zmk_hid_unregister_mod(code - LEFT_CONTROL);
+    if (code >= HID_EXT_USAGE_ID(HID_USAGE_KEY_KEYBOARD_LEFTCONTROL) &&
+        code <= HID_EXT_USAGE_ID(HID_USAGE_KEY_KEYBOARD_RIGHT_GUI)) {
+        return zmk_hid_unregister_mod(code - HID_EXT_USAGE_ID(HID_USAGE_KEY_KEYBOARD_LEFTCONTROL));
```
`HID_EXT_USAGE_ID` superfluous?

innovaker reviewed (changes requested):
Oops.

innovaker commented on docs/docs/feature/encoders.md:
````diff
@@ -25,7 +25,7 @@ Rotation is handled separately as a type of sensor. The behavior for this is set
 ```
 
-- `BINDING` is one of two rotation bindings that are currently defined, `&inc_dec_cp` for consumer key presses or `&inc_dec_kp` for normal key presses (see [Key Press](/docs/behavior/key-press) for the difference between the two).
+- `BINDING` is currently only the one implemented bindings; `&inc_dec_kp` for key presses (see [Key Press](/docs/behavior/key-press) for details on available keycodes).
````
> only the one implemented bindings
Have you had your coffee today?

petejohanson replied:
Not enough, clearly.

petejohanson replied:
Yeah, updated. Not sure if my tired brain wrote copy that's significantly better. Thoughts?

innovaker replied:
I think the approach here is to rephrase the stuff around it, e.g.

`BINDING`: the only relevant behavior currently implemented is `&inc_dec_kp`.
`CW_KEY`: the keycode ...
`CCW_KEY`: the keycode ...

innovaker replied:
Scratch that!  GitHub was showing me an outdated version.

innovaker commented on docs/docs/feature/underglow.md:
```diff
@@ -87,7 +87,7 @@ Here's an example of an nRF52 SPI definition:
 
 If you are configuring SPI for an nRF52840 (or other nRF52) based board, double check that you are using pins that aren't restricted to low frequency I/O.
-Ignoring these restrictions may result in poor wireless performance. You can find the list of low frequency I/O pins [here](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fps_nrf52840%2Fpin.html&cp=4_0_0_6_0).
+Ignoring these restrictions may result in poor wireless performance. You can find the list of low frequency I/O pins [here](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fps_nrf52840%2Fpin.html&kp=4_0_0_6_0).
```
I'm thinking you haven't.

innovaker replied:
That said, the URL still works 🥇

petejohanson replied:
Reverted. Over ambitious VS Code Search and Replace. Good catch.

innovaker reviewed (approved):
LGTM!

e2e tests:
- [e2e-2020-11-05T19-46-02-ble.log](https://github.com/zmkfirmware/zmk/files/5496641/e2e-2020-11-05T19-46-02-ble.log)
- [e2e-2020-11-05T19-46-02-usb.log](https://github.com/zmkfirmware/zmk/files/5496642/e2e-2020-11-05T19-46-02-usb.log)
